aws_lambda_powertools/utilities/idempotency/persistence/dynamodb.py

```diff
@@ -246,13 +244,20 @@ def _put_record(self, data_record: DataRecord) -> None:
                     ":now_in_millis": {"N": str(int(now.timestamp() * 1000))},
                     ":inprogress": {"S": STATUS_CONSTANTS["INPROGRESS"]},
                 },
-                **self.return_value_on_condition,  # type: ignore
+                **self.return_value_on_condition,  # type: ignore[arg-type]
             )
         except ClientError as exc:
             error_code = exc.response.get("Error", {}).get("Code")
             if error_code == "ConditionalCheckFailedException":
-                old_data_record = self._item_to_data_record(exc.response["Item"]) if "Item" in exc.response else None
-                if old_data_record is not None:
+                try:
+                    item = exc.response["Item"]  # type: ignore[typeddict-item]
+                except KeyError:
+                    logger.debug(
+                        f"Failed to put record for already existing idempotency key: {data_record.idempotency_key}",
+                    )
+                    raise IdempotencyItemAlreadyExistsError() from exc
+                else:
+                    old_data_record = self._item_to_data_record(item)
```

Comment:
What I've done here was replace the "Look Before You Leap" style by the ["Easier to Ask Forgiveness Than Permission"](https://stackoverflow.com/a/12270384/2654518), which is arguably more pythonic.

Doing:
```python
r = f(d["key"]) if "key" in d else None
if r is not None:
    # key exists
else:
    # key does not exist
```
is the same as (given the function `f` never returns None, which is the case for `self._item_to_data_record`, and `d["key"]` is never None, which is also the case for `exc.response["Item"]` because otherwise the function would fail):
```python
try:
    v = d["key"]
except KeyError:
    # key does not exist
else:
    r = f(v)
    # key exists
```
The latter also does one less dict lookup than the former. Also former code has two ifs and latter has none, which I hope you can appreciate is less complex.  :- )

Reply:
Thanks again for sharing the idea/concepts behind your code, it's absolutely amazing to educate/teach others in this area.
I was just running your e2e test to make sure everything is working as expected!!